File: packages/sdkwork-utils-go/datetime.go
package utils

import "time"

const DefaultPattern = "iso8601"

const iso8601MillisLayout = "2006-01-02T15:04:05.000Z"

func Now() time.Time {
	return time.Now().UTC()
}

func FormatDatetime(value time.Time, pattern string) string {
	if pattern == "" {
		pattern = DefaultPattern
	}
	if pattern != DefaultPattern {
		return ""
	}
	return value.UTC().Format(iso8601MillisLayout)
}

func ParseDatetime(value string, pattern string) (time.Time, bool) {
	if pattern == "" {
		pattern = DefaultPattern
	}
	if pattern != DefaultPattern {
		return time.Time{}, false
	}
	parsed, err := time.Parse(time.RFC3339Nano, value)
	if err != nil {
		parsed, err = time.Parse(time.RFC3339, value)
	}
	if err != nil {
		return time.Time{}, false
	}
	return parsed.UTC(), true
}

func AddDays(value time.Time, days int) time.Time {
	return value.Add(time.Duration(days) * 24 * time.Hour)
}

func AddHours(value time.Time, hours int) time.Time {
	return value.Add(time.Duration(hours) * time.Hour)
}

func AddMinutes(value time.Time, minutes int) time.Time {
	return value.Add(time.Duration(minutes) * time.Minute)
}

func DiffMillis(earlier, later time.Time) int64 {
	return later.Sub(earlier).Milliseconds()
}

func IsBefore(first, second time.Time) bool {
	return first.Before(second)
}

func IsAfter(first, second time.Time) bool {
	return first.After(second)
}

func StartOfDayUTC(value time.Time) time.Time {
	utc := value.UTC()
	return time.Date(utc.Year(), utc.Month(), utc.Day(), 0, 0, 0, 0, time.UTC)
}

func EndOfDayUTC(value time.Time) time.Time {
	start := StartOfDayUTC(value)
	return start.Add(24*time.Hour - time.Millisecond)
}

func ToUnixMillis(value time.Time) int64 {
	return value.UTC().UnixMilli()
}

func FromUnixMillis(value int64) (time.Time, bool) {
	return time.UnixMilli(value).UTC(), true
}

func IsSameInstant(first, second time.Time) bool {
	return first.UTC().UnixMilli() == second.UTC().UnixMilli()
}

File: packages/sdkwork-utils-rust/src/datetime.rs
use chrono::{DateTime, Duration, NaiveTime, TimeZone, Utc};

pub const DEFAULT_PATTERN: &str = "%Y-%m-%dT%H:%M:%S%.3fZ";

pub fn now() -> DateTime<Utc> {
    Utc::now()
}

pub fn format_datetime(value: DateTime<Utc>, pattern: Option<&str>) -> String {
    value.format(pattern.unwrap_or(DEFAULT_PATTERN)).to_string()
}

pub fn parse_datetime(value: &str, pattern: Option<&str>) -> Option<DateTime<Utc>> {
    let pattern = pattern.unwrap_or(DEFAULT_PATTERN);
    DateTime::parse_from_str(value, pattern)
        .ok()
        .map(|dt| dt.with_timezone(&Utc))
        .or_else(|| {
            chrono::DateTime::parse_from_rfc3339(value)
                .ok()
                .map(|dt| dt.with_timezone(&Utc))
        })
}

pub fn add_days(value: DateTime<Utc>, days: i64) -> DateTime<Utc> {
    value + Duration::days(days)
}

pub fn add_hours(value: DateTime<Utc>, hours: i64) -> DateTime<Utc> {
    value + Duration::hours(hours)
}

pub fn add_minutes(value: DateTime<Utc>, minutes: i64) -> DateTime<Utc> {
    value + Duration::minutes(minutes)
}

pub fn diff_millis(earlier: DateTime<Utc>, later: DateTime<Utc>) -> i64 {
    (later - earlier).num_milliseconds()
}

pub fn is_before(first: DateTime<Utc>, second: DateTime<Utc>) -> bool {
    first < second
}

pub fn is_after(first: DateTime<Utc>, second: DateTime<Utc>) -> bool {
    first > second
}

pub fn start_of_day_utc(value: DateTime<Utc>) -> DateTime<Utc> {
    let date = value.date_naive();
    Utc.from_utc_datetime(&date.and_time(NaiveTime::from_hms_opt(0, 0, 0).expect("midnight")))
}

pub fn end_of_day_utc(value: DateTime<Utc>) -> DateTime<Utc> {
    let date = value.date_naive();
    Utc.from_utc_datetime(
        &date.and_time(NaiveTime::from_hms_milli_opt(23, 59, 59, 999).expect("end of day")),
    )
}

pub fn to_unix_millis(value: DateTime<Utc>) -> i64 {
    value.timestamp_millis()
}

pub fn from_unix_millis(value: i64) -> Option<DateTime<Utc>> {
    DateTime::from_timestamp_millis(value)
}

pub fn is_same_instant(first: DateTime<Utc>, second: DateTime<Utc>) -> bool {
    first.timestamp_millis() == second.timestamp_millis()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_and_diff() {
        let first = parse_datetime("2024-01-01T00:00:00.000Z", None).unwrap();
        let second = add_hours(first, 2);
        assert_eq!(diff_millis(first, second), 7_200_000);
        assert!(is_before(first, second));
    }
}
